File: core/x/javaapi/api/src/main/java/viaduct/java/api/types/OffsetCursor.java
package viaduct.java.api.types;

import viaduct.tenant.runtime.support.OffsetCursorCodec;

/**
 * A cursor for offset-based pagination.
 *
 * <p>OffsetCursor wraps an encoded cursor string that represents a position in a paginated list.
 * The cursor format is: {@code Base64("__viaduct:idx:<offset>")}.
 *
 * <p>Java equivalent of Kotlin's {@code viaduct.api.types.OffsetCursor}. The encode/decode logic
 * lives in the language-neutral {@link OffsetCursorCodec}, shared with the Kotlin Tenant API, so
 * cursors produced by either language are byte-for-byte interchangeable.
 *
 * <ul>
 *   <li>{@link #getValue()} holds the <b>encoded</b> cursor string (e.g. {@code
 *       "X192aWFkdWN0OmlkeDow"})
 *   <li>{@link #toOffset()} <b>decodes</b> the cursor string to get the offset
 *   <li>{@link #fromOffset(int)} <b>creates</b> an OffsetCursor by encoding the offset
 * </ul>
 */
public final class OffsetCursor {
  private final String value;

  public OffsetCursor(String value) {
    this.value = value;
  }

  /** Returns the encoded cursor string (Base64 encoded). */
  public String getValue() {
    return value;
  }

  /**
   * Decodes this cursor to get the offset value.
   *
   * @return the decoded offset
   * @throws IllegalArgumentException if the cursor is invalid, malformed, or uses an unknown format
   */
  public int toOffset() {
    return OffsetCursorCodec.INSTANCE.decode(value);
  }

  /**
   * Creates an OffsetCursor from an offset value. Encodes the offset into the cursor string format.
   *
   * @param offset the zero-based offset position
   * @return an OffsetCursor containing the encoded cursor string
   * @throws IllegalArgumentException if offset is negative
   */
  public static OffsetCursor fromOffset(int offset) {
    return new OffsetCursor(OffsetCursorCodec.INSTANCE.encode(offset));
  }

  /**
   * Checks if a cursor string is valid without throwing.
   *
   * @param cursor the cursor string to validate
   * @return true if the cursor is valid and decodable
   */
  public static boolean isValid(String cursor) {
    return OffsetCursorCodec.INSTANCE.isValid(cursor);
  }
}

File: core/tenant/api/src/main/kotlin/viaduct/api/types/OffsetCursor.kt
package viaduct.api.types

import viaduct.apiannotations.ExperimentalApi
import viaduct.tenant.runtime.support.OffsetCursorCodec

/**
 * A cursor for offset-based pagination.
 *
 * OffsetCursor wraps an encoded cursor string that represents a position
 * in a paginated list. The cursor format is: Base64("__viaduct:idx:<offset>")
 *
 * Key Design:
 * - `value: String` holds the **encoded cursor string** (e.g., "X192aWFkdWN0OmlkeDow")
 * - `toOffset()` **decodes** the cursor string to get the offset Int
 * - `fromOffset(offset)` **creates** an OffsetCursor by encoding the offset
 *
 * The encode/decode logic lives in the language-neutral [OffsetCursorCodec] so the Kotlin and Java
 * Tenant APIs produce byte-identical cursors; this value class is a thin typed wrapper over it.
 *
 * Usage:
 * ```kotlin
 * // Create cursor from offset
 * val cursor = OffsetCursor.fromOffset(42)
 *
 * // Get the encoded string value for GraphQL response
 * val cursorString: String = cursor.value  // "X192aWFkdWN0OmlkeDo0Mg"
 *
 * // Decode cursor back to offset
 * val offset: Int = cursor.toOffset()  // 42
 *
 * // From incoming cursor string
 * val incomingCursor = OffsetCursor(cursorString)
 * val decodedOffset = incomingCursor.toOffset()
 * ```
 *
 * @property value The encoded cursor string (Base64 encoded)
 */
@ExperimentalApi
@JvmInline
value class OffsetCursor(val value: String) {
    /**
     * Decode this cursor to get the offset value.
     *
     * @return The decoded offset
     * @throws IllegalArgumentException if cursor is invalid, malformed, or uses unknown format
     */
    fun toOffset(): Int = OffsetCursorCodec.decode(value)

    companion object {
        /**
         * Create an OffsetCursor from an offset value.
         * Encodes the offset into the cursor string format.
         *
         * @param offset The zero-based offset position
         * @return An OffsetCursor containing the encoded cursor string
         * @throws IllegalArgumentException if offset is negative
         */
        @ExperimentalApi
        fun fromOffset(offset: Int): OffsetCursor = OffsetCursor(OffsetCursorCodec.encode(offset))

        /**
         * Check if a cursor string is valid without throwing.
         *
         * @param cursor The cursor string to validate
         * @return true if cursor is valid and decodable
         */
        @ExperimentalApi
        fun isValid(cursor: String): Boolean = OffsetCursorCodec.isValid(cursor)
    }
}
